mp/src/game/server/momentum/mapzones_build.cpp

```diff
@@ -0,0 +1,978 @@
+}
+
+void CMomPointZoneBuilder::FixPointOrder(CUtlVector<Vector> &points)
+{
+    // Correct the points to be in clockwise order
+
+    int len = points.Count();
+    if (len < 3)
+        return;
+
+
+    // Add all the dot products together.
+    // This magically figures out which way we're going
+    float total_dot = 0.0f;
+
+    for (int i = 0; i < len; i++)
+    {
+        int inext = (i+1 >= len) ? 0 : i+1;
+        int iprev = (i-1 < 0) ? len-1 : i-1;
+
+        auto& next = points[inext].AsVector2D();
+        auto& prev = points[iprev].AsVector2D();
+        auto& cur = points[i].AsVector2D();
+
+        auto n = next - cur;
+        n.NormalizeInPlace();
+        auto p = prev - cur;
+        p.NormalizeInPlace();
+
+        auto pleft = Vector2D(-p.y, p.x);
+
+
+
+        auto pdot = n.Dot(pleft);
+        total_dot += pdot;
+    }
+
+    if (total_dot >= 0.0f)
+        return;
+
+
+    // Switch em up.
+
+    DevMsg("Fixing points to clockwise order...\n");
+        
+    Vector temp;
+    for (int i = 0; i < (len/2); i++)
+    {
+        int n = len-i-1;
+
+        temp = points[n];
+        points[n] = points[i];
+        points[i] = temp;
+    }
+}
+
+bool CMomPointZoneBuilder::BuildZone(CBasePlayer *pPlayer, const Vector *vecAim)
+{
+    if (!IsReady())
+    {
+        Warning("Not enough points to build a convex shape!\n");
+        return false;
+    }
+
+
+    if (pPlayer)
+    {
+        if (!m_bGetHeight)
+        {
+            m_bGetHeight = true;
+            return false;
+        }
+
+        m_flHeight = CMomBoxZoneBuilder::GetZoneHeightToPlayer(pPlayer, m_vPoints[0]);
+        m_bGetHeight = false;
+    }
+
+
+    const float flDefHeight = 128.0f;
+    const float epsilon = 0.1f;
+    if (fabsf(m_flHeight) < epsilon)
+    {
+        m_flHeight = flDefHeight;
+    }
+
+
+    // Get the points relative to our center
+    CUtlVector<Vector> relpoints(m_vPoints.Count(), m_vPoints.Count());
+    relpoints.SetSize(m_vPoints.Count());
+    Init(relpoints);
+
+
+    // Divide into convex shapes
+    CMomHulls_t hulls;
+    Decompose(relpoints, hulls);
+    
+
+    
+    int nHulls = hulls.Count();
+
+    Assert(nHulls >= 1);
+
+    if (!nHulls)
+    {
+        Warning("No convex hulls were found from points!\n");
+        return false;
+    }
+
+    DrawDebugLines(hulls);
+
+
+    pvertarray_t **ppHulls = nullptr;
+    // Build the actual three dimensional vertex arrays
+    BuildVertArray(hulls, &ppHulls, nHulls);
+
+    m_pPhysCollide = BuildPhysCollide(ppHulls, nHulls);
+
+
+    hulls.PurgeAndDeleteElements();
+    delete[] ppHulls;
+
+
+    return m_pPhysCollide != nullptr;
+}
+
+bool CMomPointZoneBuilder::IsReady() const
+{
+    return m_vPoints.Count() >= 3;
+}
+
+bool CMomPointZoneBuilder::IsDone() const
+{
+    return m_pPhysCollide != nullptr;
+}
+
+void CMomPointZoneBuilder::Reset()
+{
+    ResetMe();
+}
+
+void CMomPointZoneBuilder::ResetMe()
+{
+    m_pPhysCollide = nullptr;
```

Comment:
Don't know if its actually needed, but it might be a good idea to call `physcollision->DestroyCollide(m_pPhysCollide)` here to avoid being leaky.